File: backend/AutoAnders/src/main/java/Gloyoo/AutoAnders/Cars/entity/Car.java
package Gloyoo.AutoAnders.Cars.entity;

import Gloyoo.AutoAnders.CarPictures.entity.CarPicture;
import Gloyoo.AutoAnders.user.entity.User;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Entity
@Table(name = "cars",
        uniqueConstraints = {
                @UniqueConstraint(name = "uk_cars_license_plate", columnNames = "license_plate")
        }
)

public class Car {
    @GeneratedValue(strategy = GenerationType.UUID)
    @Id
    UUID id;


    private String title;
    private String subtitle;

    private Integer yearOfManufacture;
    private Integer mileage;
    private String power;
    private String referenceNumber;
    private BigDecimal price;

    private LocalDate firstRegistrationDate;
    private Integer numberOfDoors;
    private Integer wheelbase;
    private Integer numberOfCylinders;
    private String motorVehicleTax;
    private LocalDate modelDateFrom;
    private LocalDate modelDateTo;
    private Integer maxTowingWeight;
    private Integer maxTowingWeightUnbraked;
    private BigDecimal urbanFuelConsumption;
    private BigDecimal combinedFuelConsumption;
    private BigDecimal motorwayFuelConsumption;
    private Integer co2Emissions;
    private Boolean taxDeductible;
    private String chassisNumber;
    private Integer numberOfKeys;

    @Column(name = "license_plate", unique = true)
    private String licensePlate;
    private Integer engineDisplacement;
    private String colour;
    private Integer emptyWeight;
    private Integer taxAdditionPercentage;
    private String apkMotDate;
    private Boolean serviceDocumentation;
    private String location;

    private BigDecimal financialLeasePricePerMonth;
    private BigDecimal leasePrice60Months;
    private BigDecimal leasePrice48Months;
    private BigDecimal leasePrice36Months;

    @Enumerated(EnumType.STRING)
    private BodyType bodyType;

    @Enumerated(EnumType.STRING)
    private Gearbox gearbox;

    @Enumerated(EnumType.STRING)
    private Fuel fuel;

    @Enumerated(EnumType.STRING)
    private EmissionClass emissionClass;

    @Enumerated(EnumType.STRING)
    private EnergyLabel energyLabel;

    @Enumerated(EnumType.STRING)
    private PaintType paintType;

    @Enumerated(EnumType.STRING)
    private Upholstery upholstery;

    @Enumerated(EnumType.STRING)
    private Status status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @OneToMany(mappedBy = "car", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<CarPicture> pictures = new ArrayList<>();



}

File: frontend/src/app/core/interfaces/Car.ts
import { AuthUser } from "./AuthUser";


export interface Car {
    id: string;

    brand: string;
    model: string;

    title: string;
    subtitle: string;

    yearOfManufacture: number;
    mileage: number;
    power: string;
    referenceNumber: string;
    price: number;

    firstRegistrationDate: string;
    numberOfDoors: number;
    wheelbase: number;
    numberOfCylinders: number;
    motorVehicleTax: string;
    modelDateFrom: string;
    modelDateTo: string;
    maxTowingWeight: number;
    maxTowingWeightUnbraked: number;
    urbanFuelConsumption: number;
    combinedFuelConsumption: number;
    motorwayFuelConsumption: number;
    co2Emissions: number;
    taxDeductible: boolean;
    chassisNumber: string;
    numberOfKeys: number;

    licensePlate: string;
    engineDisplacement: number;
    colour: string;
    emptyWeight: number;
    taxAdditionPercentage: number;
    apkMotDate: string;
    serviceDocumentation: boolean;
    location: string;

    financialLeasePricePerMonth: number;
    leasePrice60Months: number;
    leasePrice48Months: number;
    leasePrice36Months: number;

    bodyType: BodyType;
    gearbox: Gearbox;
    fuel: Fuel;
    emissionClass: EmissionClass;
    energyLabel: EnergyLabel;
    paintType: PaintType;
    upholstery: Upholstery;
    status: CarStatus;

    user: AuthUser;
    pictures: CarPicture[];
}

export type BodyType = 'MPV' | 'SUV' | 'SEDAN' | 'HATCHBACK' | 'STATION_WAGON' | 'COUPE' | 'CABRIOLET' | 'VAN';
export type EmissionClass = 'EURO_1' | 'EURO_2' | 'EURO_3' | 'EURO_4' | 'EURO_5' | 'EURO_6';
export type EnergyLabel = 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'G';
export type Fuel = 'PETROL' | 'DIESEL' | 'ELECTRIC' | 'HYBRID' | 'LPG' | 'CNG';
export type Gearbox = 'MANUAL' | 'AUTOMATIC' | 'SEMI_AUTOMATIC';
export type PaintType = 'BASIC' | 'METALLIC' | 'PEARL' | 'MATTE';
export type CarStatus = 'Available' | 'Pending_Confirmation' | 'Booked' | 'Cancelled';
export type Upholstery = 'FABRIC' | 'LEATHER' | 'PART_LEATHER' | 'ALCANTARA';

export interface CarPicture {
    id: string;
    storage_path: string;
    title: string;
    description: string;
    width: number;
    height: number;
}

export interface AutoCatalogueLabels {
    vehiclesFound: string;
    searchPlaceholder: string;
    allBrands: string;
    allTransmissions: string;
    firstRegistrationDate: string;
    maxMileage: string;
    maxPrice: string;
    allColours: string;
    allFuels: string;
    engineCapacity: string;
    allDoorCounts: string;
    allConditions: string;
    allGearCounts: string;
    allVatOptions: string;
    allVehicles: string;
    searchButton: string;
    resetButton: string;
    viewDetails: string;
    moreButton: string;
}

export interface CarPictureRequest {
    file: File;
    title?: string;
    description?: string;
    width: number;
    height: number;
}
export type CarRequest = Omit<Car, 'id' | 'user' | 'pictures'>;
